### DISCUSSION TITLE
Getting error trying to build an image from Dockerfile

### DISCUSSION THREAD

**Discussion by pfamili081** [valhalla/valhalla#4386] [Q&A]
I cloned the repo and trying to build the image from Dockerfile and getting the below error: 
ERROR: failed to solve: process "/bin/sh -c bash /usr/local/src/valhalla/scripts/install-linux-deps.sh" did not complete successfully: exit code: 100
Would appreciate your help with this.

**Comment by kevinkreiser** [comment]
you need to show us the commands you ran and a bit more of the output

**Reply by pfamili081** [reply]
docker build C:\Users\PFAMILI\Documents\GitHub\valhalla --network=host
......
Dockerfile:24
22 | WORKDIR /usr/local/src/valhalla
23 | COPY ./scripts/install-linux-deps.sh /usr/local/src/valhalla/scripts/install-linux-deps.sh
24 | >>> RUN bash /usr/local/src/valhalla/scripts/install-linux-deps.sh
25 | RUN rm -rf /var/lib/apt/lists/*
26 |

**Reply by pfamili081** [reply]
--------------------
ERROR: failed to solve: process "/bin/sh -c bash /usr/local/src/valhalla/scripts/install-linux-deps.sh" did not complete successfully: exit code: 100

**Reply by kevinkreiser** [reply]
the docs tell you why this wont work:

https://docs.docker.com/network/drivers/host/

> The host networking driver only works on Linux hosts, and is not supported on Docker Desktop for Mac, Docker Desktop for Windows, or Docker EE for Windows Server.

it looks like you are on windows and in order to build the code we need the internet so that we can download dependencies inside the docker container. if that `--network=host` parameter disables networking because it doesnt work on windows that would explain why the dependency installer script fails. i cant test this because im not on windows though. try removing the `--network` flag from your command

**Reply by pfamili081** [reply]
didn't change the error

**Reply by kevinkreiser** [reply]
there should be more output from that script. failed to solve sounds more like failed REsolve meaning find the script. can you try bulding from the root of the repository?

**Reply by pfamili081** [reply]
The final goal was to use map matching engine in GCP, along with tiles (I uploaded a sample for now in my bucket) and my GPS points (I have tables in BQ). After lots of research I believe "Cloud Run" is not a good match for me to do that. I was wondering if you happen to know how can I use this package in GCP? Thank you for your help.

**Reply by ImreSamu** [reply]
> how can I use this package in GCP? 

I am currently operating several Valhalla Docker services on Google Compute Engine, using n2d instances equipped with Ubuntu. Additionally, both Docker Compose and Docker have been installed. The setup is straightforward.

If you are working on Windows:
- Kindly use WSL2 (Windows Subsystem for Linux 2) with the latest version of Ubuntu.
- For building Docker images, please avoid using the Valhalla ZIP download. This method sometimes fails to maintain the necessary permissions for the files. ( see [issue](https://github.com/actions/upload-artifact/issues/38) ) Instead, kindly use the `git clone https://github.com/valhalla/valhalla.git` command to obtain the files.

**Reply by pfamili081** [reply]
Thank you for your reply. A question; I am following "https://ikespand.github.io/posts/meili/" and have my notebook and result based on this instruction. Do you happen to check "https://github.com/gis-ops/docker-valhalla", and any experience to have all the steps from that in GCP. Would really appreciate your help in that.

**Reply by ImreSamu** [reply]
The andorra example ( from https://github.com/gis-ops/docker-valhalla ) looks OK ( on my ubuntu + AMD64 )
```
# download a file to custom_files and start valhalla
mkdir custom_files
wget -O custom_files/andorra-latest.osm.pbf https://download.geofabrik.de/europe/andorra-latest.osm.pbf
docker run -dt --name valhalla_gis-ops -p 8002:8002 -v $PWD/custom_files:/custom_files ghcr.io/gis-ops/docker-valhalla/valhalla:latest

# test:
$ curl 127.0.0.1:8002/status | jq .
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   238  100   238    0     0  93516      0 --:--:-- --:--:-- --:--:--  116k
{
  "version": "3.4.0",
  "tileset_last_modified": 1699754588,
  "available_actions": [
    "status",
    "centroid",
    "expansion",
    "transit_available",
    "trace_attributes",
    "trace_route",
    "isochrone",
    "optimized_route",
    "sources_to_targets",
    "height",
    "route",
    "locate"
  ]
}

```

tested now ; 
and my docker log :  https://gist.github.com/ImreSamu/6ee849c6f179e490b21df432592e5bdd
please first test with smaller area  like:  andorra and see the log;
check the log and wait for the preprocessing ( "INFO: Successfully built files" ) ! 

```
....
INFO: Successfully built files: /custom_files/andorra-latest.osm.pbf
INFO: Hashing files: /custom_files/andorra-latest.osm.pbf
INFO: Hashing file: /custom_files/andorra-latest.osm.pbf
e624de2cd88a7686a313d34043ea8bbed43999c16d999e9624c652a45080b367
2023-11-12 02:03:08,120  INFO: Finished tarring 7 tiles to /custom_files/valhalla_tiles.tar
WARNING: User valhalla is running with sudo privileges. Try building the image with a host user's UID & GID.
INFO: Found config file. Starting valhalla service!
2023/11/12 02:03:08.225426 [INFO] Tile extract successfully loaded with tile count: 7
2023/11/12 02:03:08.225464 [INFO] Tile extract successfully loaded with tile count: 7
....
```

**Reply by pfamili081** [reply]
Thank you for the detailed reply. I could run it too. I have 2 questions, could we tag the image created from this run and push it to artifact registry? and if yes (which I did but not sure if it works or not because the port is 8002 but the GCP port is 8080) could we deploy it from there in GKE? if the answer is yes too, the last part is could we use managed notebook from Vertex AI to send request to the artifact registry as a kernel? Thank you so much for your answers.

**Reply by ImreSamu** [reply]
> could we tag the image created from this run and push it to artifact registry?

Certainly. However, the Docker image has not changed. The data necessary for Valhalla's operation has been placed in a data volume or directory. If you do not assign the data (either as a data volume or using ` -v $PWD/custom_files:/custom_files` ), it will not function correctly. I recommend that you first familiarize yourself locally with the Dockerized Valhalla and Docker.
( check the contents of the `./custom_files` directory ! )

Regarding your question about deploying it in GKE (Google Kubernetes Engine), it's likely possible if you correctly assign a data volume and transfer the necessary data there. However, I'm not able to assist with this. You should seek help from a Kubernetes expert.

**Reply by ImreSamu** [reply]
in the `./custom_files` directory ,  you can see : 
```
-rw-rw-r-- 1 Xuser Xuser 2550516 Nov 11 03:24 andorra-latest.osm.pbf  
-rw-rw-r-- 1 root  root      544 Nov 12 03:03 duplicateways.txt
-rw-rw-r-- 1 root  root       65 Nov 12 03:03 file_hashes.txt
-rw-rw-r-- 1 root  root     7497 Nov 12 03:03 valhalla.json    <---  this is the json config file, please check ! 
drwxrwxr-x 5 root  root     4096 Nov 12 03:03 valhalla_tiles  
-rw-rw-r-- 1 root  root  2529280 Nov 12 03:03 valhalla_tiles.tar  <--- this is the compressed routing tiles.
```

**Reply by ImreSamu** [reply]
for Kubernetes - please check : https://github.com/valhalla/valhalla/issues/3801
* https://github.com/itayankri/valhalla-operator
* https://medium.com/@itay.ankri/deploying-valhalla-routing-engine-on-kubernetes-using-valhalla-operator-2426e79ac746

**Reply by pfamili081** [reply]
Thank you so much.
I am kinda new to this , be appreciate if you could please explain in more detail; using  -v $PWD/custom_files:/custom_files in my push command to artifact repo? or how can we mount data volume to artifact? I only saw it for Cloud Run Service.

Do you happen to have experience to connect your Vertext AI notebook to this docker in Artifact? I read the instruction and it needed to modify the Dockerfile, and use it as a kernel;  was wondering if you happen to know another way to connect a notebook to the docker image and send request. (Looking for another way because I got error when trying to build this image with Dockerfile after cloning it, I am in Win)

**Reply by ImreSamu** [reply]
Unfortunately, I can't help much with the details. 
You will find many Docker tutorials online, and it will take about 1-2 weeks to become familiar with the technology.

However, if running the Valhalla Docker service is complicated, I suggest you don't force it. 
Here's an alternative suggestion:
1.) Try generating the `./custom_files/valhalla_tiles.tar` somewhere in the cloud.
2.) And try out Python `pyvalhalla`. (You will need `valhalla_tiles.tar` for this.)
* `pip install pyvalhalla`
* binary wheels exists for Win64, MacOS (Intel) and x86_64 Linux.

see: https://github.com/gis-ops/pyvalhalla#usage

**Comment by pfamili081** [comment]
Really appreciate your help.

Thank you so much,

Pooneh

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
